### PR TITLE
[PT-5198] Workspace singleton

### PR DESCRIPTION
**Pull request by andher1802**
Described [Here
](https://up42.atlassian.net/browse/PT-5054?atlOrigin=eyJpIjoiMDg4ZjdjNGE1NTkyNGI1NzhjYTYyZjk0Mzg1YTNkYWUiLCJwIjoiaiJ9)

Adding workspace mixing to implement active record pattern

Items:
* [x] Implemented (new) unit tests
* [ ] Updated [documentation](sdk.up42.com)

For release:
* [x] Bumped version
* [x] Added changelog

### CODE REVIEW

File: up42/auth.py [hunk 1]
```diff
@@ -51,10 +52,10 @@ def __init__(
             username: The username for the UP42 account (email UP42 console).
             password: Password for the UP42 console login.
         """
-        self.workspace_id: Optional[str] = None
+        # self.workspace_id: Optional[str] = None
```
Comment (andher1802) [inline up42/auth.py:55]: commented to mark the code to remove

File: up42/main.py [hunk 2]
```diff
@@ -11,7 +11,52 @@
...
+            cfg_file=cfg_file,
+            username=username,
+            password=password,
+        )
+
+
+workspace = __Workspace()
 
 
 def authenticate(
```
Comment (javidq) [inline up42/main.py:62]: Not needed anymore, workspace.authenticate can be exported directly

File: up42/main.py [hunk 3]
```diff
@@ -29,18 +74,11 @@ def authenticate(
...
-    global _auth
-    _auth = up42_auth.Auth(
-        cfg_file=cfg_file,
-        username=username,
-        password=password,
-    )
+    workspace.authenticate(cfg_file=cfg_file, username=username, password=password)
 
 
 def get_auth_safely() -> up42_auth.Auth:
```
Comment (javidq) [inline up42/main.py:80]: Not needed, workspace.auth to be used directly

File: up42/main.py [hunk 4]
```diff
@@ -11,7 +11,52 @@
...
+
+
+class __Workspace:  # pylint: disable=invalid-name
+    _auth: Optional[up42_auth.Auth] = None
+
+    @property
+    def auth(self):
+        if self._auth:
+            return self._auth
+        raise ValueError("User not authenticated.")
```
Comment (javidq) [inline up42/main.py:24]: Let's have a modeled error for this

File: up42/main.py [hunk 5]
```diff
@@ -11,7 +11,52 @@
...
+
+    @auth.setter
+    def auth(self, value):
+        self._auth = value
+
+    @property
+    def workspace_id(self):
```
Comment (javidq) [inline up42/main.py:31]: should be set in authenticate

File: up42/main.py [hunk 6]
```diff
@@ -51,7 +89,7 @@ def check_auth(func):
 
     @functools.wraps(func)  # required for mkdocstrings
     def inner(*args, **kwargs):
-        if _auth is None:
+        if workspace.auth is None:
             raise RuntimeError("Not authenticated, call up42.authenticate() first")
```
Comment (javidq) [inline up42/main.py:93]: This part is not needed anymore, since it will throw error when calling auth anyway

**Review by javidq — CHANGES_REQUESTED** [review]

File: up42/order.py [hunk 7]
```diff
@@ -140,7 +141,7 @@ def place(cls, auth: up42_auth.Auth, order_parameters: dict) -> "Order":
         Returns:
             Order: The placed order.
         """
-        url = host.endpoint(f"/v2/orders?workspaceId={auth.workspace_id}")
+        url = host.endpoint(f"/v2/orders?workspaceId={up42.main.workspace.workspace_id}")
```
Comment (andher1802) [inline up42/order.py:144]: Is this expected, to call workspace_id now from the workspace singleton, right? Just wanted to confirm.

Comment (javidq) [reply up42/order.py:144]: No, we need to add workspace_id to function parameters

File: up42/auth.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 """
 UP42 authentication mechanism and base requests functionality
 """
+
```
Comment (javidq) [inline up42/auth.py:4]: Not needed

File: tests/test_main.py [hunk 9]
```diff
@@ -18,9 +18,9 @@ def test_get_credits_balance():
 
 
 def test_fails_to_get_auth_safely_if_unauthenticated():
-    main._auth = None  # pylint: disable=protected-access
+    main.workspace.auth = None
     with pytest.raises(ValueError):
```
Comment (javidq) [inline tests/test_main.py:22]: It has a specific error now

File: tests/test_main.py [hunk 10]
```diff
@@ -7,7 +7,7 @@
 
 @pytest.fixture(autouse=True)
 def setup_auth_mock(auth_mock):
```
Comment (javidq) [inline tests/test_main.py:9]: Do we still need this?

**Review by javidq — CHANGES_REQUESTED** [review]

File: up42/__init__.py [hunk 11]
```diff
@@ -57,10 +57,10 @@
         initialize_order,
         initialize_asset,
         initialize_webhook,
-        authenticate,
         get_webhooks,
         create_webhook,
         get_webhook_events,
         get_credits_balance,
     ]
+    if hasattr(obj, "__name__")
```
[the remark this answers is not in the document]
Comment (andher1802) [reply up42/__init__.py:66]: Pylint is triggering an error if we do not add this. Alternative option is to set an ignore. :)

File: up42/auth.py (hunk 8 above)
Comment (andher1802) [reply up42/auth.py:4]: I did not get this one, what is not needed? Are you referring to the space added?
It is automatically added by the linter in vscode.

File: tests/test_main.py (hunk 10 above)
Comment (andher1802) [reply tests/test_main.py]: This was needed for overcoming the pylint unused argument warning, I added as an ignore line.

File: tests/fixtures/fixtures_auth.py [hunk 12]
```diff
@@ -24,4 +25,5 @@ def auth_mock(requests_mock: req_mock.Mocker) -> up42_auth.Auth:
         url=url_get_credits_balance,
         json=constants.JSON_BALANCE,
     )
-    return up42_auth.Auth(username=constants.USER_EMAIL, password=constants.PASSWORD)
```
Comment (javidq) [inline tests/fixtures/fixtures_auth.py:27]: Why is this change needed?

File: CHANGELOG.md [hunk 13]
```diff
@@ -29,6 +29,13 @@ You can check your current version with the following command:
     ```
 
 For more information, see [UP42 Python package description](https://pypi.org/project/up42-py/).
+## 1.1.0a1
+
+**May 2, 2024**
+
+- Adding __Workspace hidden class and worskspace instance as a singleton class in `main.py`
```
Comment (javidq) [inline CHANGELOG.md:36]: Typo

File: tests/test_auth.py [hunk 14]
```diff
@@ -96,7 +96,6 @@ def create_auth(requests_mock: req_mock.Mocker):
 class TestAuth:
```
Comment (javidq) [inline tests/test_auth.py:96]: There's more to clean up here, workspace endpoint mocking is not needed any more

File: tests/test_initialization.py [hunk 15]
```diff
@@ -7,15 +7,15 @@
 
 
 def test_initialize_object_without_auth_raises():
-    main._auth = None  # pylint: disable=protected-access
+    main.workspace.auth = None  # pylint: disable=protected-access
```
Comment (javidq) [inline tests/test_initialization.py:10]: There's no protected access anymore

File: tests/test_main.py [hunk 16]
```diff
@@ -5,25 +5,19 @@
 from .fixtures import fixtures_globals as constants
 
 
-@pytest.fixture(autouse=True)
```
Comment (javidq) [inline tests/test_main.py:8]: Let's leave this autouse fixture here to avoid unused auth_mock every where else

File: tests/test_main.py [hunk 17]
```diff
@@ -5,25 +5,19 @@
...
-
-
-def test_get_credits_balance():
+def test_get_credits_balance(auth_mock):  # pylint: disable=unused-argument
     balance = main.get_credits_balance()
     assert isinstance(balance, dict)
     assert "balance" in balance
 
 
 def test_fails_to_get_auth_safely_if_unauthenticated():
```
Comment (javidq) [inline tests/test_main.py:14]: to be renamed since there's no such method anymore

File: tests/test_storage.py [hunk 18]
```diff
@@ -272,7 +272,7 @@ def test_get_orders_v2_endpoint_params(auth_mock, requests_mock, params, expecte
     url_params = "&".join(
         [
             "sort=createdAt%2Cdesc",
-            f"workspaceId={constants.WORKSPACE_ID}" if params["workspace_orders"] else "",
+            (f"workspaceId={constants.WORKSPACE_ID}" if params["workspace_orders"] else ""),
```
Comment (javidq) [inline tests/test_storage.py:275]: Why do we need this change?

File: up42/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (javidq) [reply up42/__init__.py:66]: It means something is off, otherwise we would get the warning before as well

File: up42/catalog.py [hunk 19]
```diff
@@ -11,6 +11,7 @@
 import tqdm
 from shapely import geometry as geom  # type: ignore
 
+import up42.main
```
Comment (javidq) [inline up42/catalog.py:14]: from up42 import main

File: up42/catalog.py [hunk 20]
```diff
@@ -150,7 +151,7 @@ def place_order(
             warnings.warn(message, DeprecationWarning, stacklevel=2)
         elif order_parameters is None:
             raise ValueError("Please provide the 'order_parameters' parameter!")
-        placed_order = order.Order.place(self.auth, order_parameters)  # type: ignore
+        placed_order = order.Order.place(self.auth, order_parameters, up42.main.workspace.workspace_id)  # type: ignore
```
Comment (javidq) [inline up42/catalog.py:154]: Why do we need type:ignore here?

File: up42/storage.py [hunk 21]
```diff
@@ -39,9 +38,9 @@ class Storage:
     ```
     """
 
-    def __init__(self, auth: up42_auth.Auth):
+    def __init__(self, auth: up42.Auth):
```
Comment (javidq) [inline up42/storage.py:41]: Let's keep up42_auth and avoid using exported namespace

File: up42/storage.py [hunk 22]
```diff
@@ -8,9 +8,8 @@
 import geopandas  # type: ignore
 from shapely import geometry as shp_geometry  # type: ignore
 
-from up42 import asset, asset_searcher
-from up42 import auth as up42_auth
-from up42 import host, order, stac_client, utils
+import up42.main
```
Comment (javidq) [inline up42/storage.py:11]: from up42 import main

File: up42/webhooks.py [hunk 23]
```diff
@@ -1,5 +1,6 @@
 from typing import List, Optional
 
+import up42.main
```
Comment (javidq) [inline up42/webhooks.py:3]: from up42 import main

File: up42/main.py [hunk 24]
```diff
@@ -11,54 +10,55 @@
...
-    """
-    global _auth
-    _auth = up42_auth.Auth(
-        cfg_file=cfg_file,
-        username=username,
-        password=password,
-    )
+class __Workspace:  # pylint: disable=invalid-name
+    _auth: Optional[up42_auth.Auth] = None
+    workspace_id: Optional[str] = None
```
Comment (javidq) [inline up42/main.py:20]: can we name it `id`?

File: tests/test_main.py [hunk 25]
```diff
@@ -5,25 +5,19 @@
 from .fixtures import fixtures_globals as constants
```
Comment (javidq) [inline tests/test_main.py:6]: Test coverage is missing for new functionality

File: up42/main.py [hunk 26]
```diff
@@ -11,54 +10,55 @@
...
-        username: The username for the UP42 account (email UP42 console).
-        password: Password for the UP42 console login.
-    """
-    global _auth
-    _auth = up42_auth.Auth(
-        cfg_file=cfg_file,
-        username=username,
-        password=password,
-    )
+class __Workspace:  # pylint: disable=invalid-name
```
Comment (javidq) [inline up42/main.py:18]: Let's rename it to _Workspace instead of disabling pylint

**Review by javidq — CHANGES_REQUESTED** [review]

**Review by javidq — CHANGES_REQUESTED** [review]

File: tests/fixtures/fixtures_auth.py (hunk 12 above)
Comment (andher1802) [reply tests/fixtures/fixtures_auth.py:27]: Because returning only Auth is enough for tests that uses only the `auth.Auth` class, but not for the ones that also use workspace. e.g. test_order_from_catalog that uses the workspace mock.

Screenshot 2024-05-22 at 13 59 34

File: up42/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (andher1802) [reply up42/__init__.py]: After some research, I finally found what is off here.

We have changed the `authenticate` variable to store the method of a Class. It seems that before `mypy` was able to infer that there were Classes, and functions in the list. Now, we have a third type that is variable, therefore we have in that list Classes, Functions, and one variable that store a method of a Class.

Since the list contains a mix of classes, functions, and variables, `mypy` infers the most general type that can accommodate all elements, which is likely `object`. The object type in Python doesn't have a __name__ attribute, leading to the error.

I figure out that splitting the assingment for Classes, and functions makes `mypy` happy and infers correctly the type. I also tested this and it works as well: 

```
__all__ = [
    Asset.__name__,
    Auth.__name__,
    Catalog.__name__,
    Order.__name__,
    Storage.__name__,
    Tasking.__name__,
    Webhook.__name__,
    get_example_aoi.__name__,
    read_vector_file.__name__,
    initialize_catalog.__name__,
    initialize_tasking.__name__,
    initialize_storage.__name__,
    initialize_order.__name__,
    initialize_asset.__name__,
    initialize_webhook.__name__,
    authenticate.__name__,
    get_webhooks.__name__,
    create_webhook.__name__,
    get_webhook_events.__name__,
    get_credits_balance.__name__,
] 
```
We can think on a better solution, perharps assinging `authenticate` to a better type or typing better, I did some test but I ended with splitting the assignment of the __all__ property to Classes and functions.

File: tests/test_storage.py (hunk 18 above)
Comment (andher1802) [reply tests/test_storage.py:275]: no needed, autoformat from my IDE

**Review by javidq — APPROVED** [review]
